Geometric mesh data added for Damaris in-situ visualisation

jcbowden opened the pull request:
Added access to DUNE mesh geometry and passing through data to Damaris; 

Updated flow command line options so users can specify Python or Paraview script names and other parameters that control Damaris:
- Simulation name
- Number of dedicated cores or dedicated nodes
- Shared memory region size
- switch to turn off HDF5 output.
- Damaris logging level

jcbowden commented:
It would be great if someone could just get jenkins to build this serial, as the tests are useful and best to know now.

akva2 commented:
jenkins build this serial please

jcbowden commented:
Well, I don't understand the test results - they say all tests passed, however the command line output shows a build failure. Can someone explain this to me please?

akva2 commented:
Everything passed in the serial build since there is no damaris. In parallel it failed to build, with enabled damaris.

jcbowden commented:
There seems to be an issue with the command line arguments I have added. 

I cannot reproduce the error on my local machine. I updated to the latest masters. Could I get the CMake configuration that is used on the Jenkins server?

akva2 commented:
see the PR i sent to your branch.

jcbowden commented:
Thanks for the direction with the command line arguments (and the other code tidy-up). My build has HAVE_DAMARIS defined in config.h so it definitely has Damaris enabled.  Still no issue found. I did not have the headers added in CMakeLists_files.cmake:
  list (APPEND PUBLIC_HEADER_FILES opm/simulators/utils/DamarisOutputModule.hpp)
  list (APPEND PUBLIC_HEADER_FILES opm/simulators/utils/DamarisKeywords.hpp)
This may have been an problem?
Anyway, i'll adapt my branch with your suggestions and try again.

jcbowden commented:
Could this bee tested again please?

akva2 commented:
jenkins build this serial please

jcbowden commented:
sorry, i'm not sure how that issue slipped through. I'll try again soon.

jcbowden commented:
And again, thanks.

akva2 commented:
jenkins build this serial please

jcbowden commented:
I'm really pushing this test driven development bandwagon today. Could you please ask Jenkins again.

akva2 commented:
jenkins build this serial please

jcbowden commented:
So, this version passes, however, there is an issue with an unused parameter - it is only unused as neither Paraview or Python have been enabled in the Damaris library used for testing.

bska commented:
> there is an issue with an unused parameter

Right.  The typical remedy for that is to tag the parameter as [`[[maybe_unused]]`](https://en.cppreference.com/w/cpp/language/attributes/maybe_unused).

bska commented:
jenkins build this serial please

jcbowden commented:
Is there further review of this PR planned? Should I rebase first?

jcbowden commented on opm/simulators/utils/DamarisOutputModule.cpp:
```diff
@@ -30,48 +29,51 @@
-    }
-
-    int damaris_err;
+    int dam_err_;
```
I have an issue with this as when I search for the Damaris API functions, "grep "damaris_ ." then I get lines with the error variable, whichmakes output unclear.

jcbowden commented on opm/simulators/utils/DamarisKeywords.cpp:
```diff
@@ -17,55 +17,210 @@
+{
+    // From c++17  : std::filesystem::exists(filename_in);
+    
+    int retint = 0 ;
```
It is my habit, I should change it for the OPM code, so will search and replace.

atgeirr commented on opm/simulators/utils/DamarisVar.hpp:
```diff
@@ -0,0 +1,620 @@
+                }
+            }
+
+            void SetPointersToDamarisShmem( T ** ptr_in )
```
What is the use case for this overload? Can it be removed? I think an outside caller could just get the data pointer using the accessor if necessary.

atgeirr reviewed (commented):
Done with the review, I think this should be merged once the noted issues are resolved. Nothing really serious or hard to fix I hope.

jcbowden commented on opm/simulators/utils/DamarisOutputModule.cpp:
```diff
@@ -30,48 +29,51 @@
+        if (mpiRank == 0) {
+            myMod.SaveXMLStringToFile(damaris_xml_filename_str);
+        }
+        std::cout << "INFO: Initializing Damaris using internally built file:" << damaris_xml_filename_str << " (N.B. use FLOW_DAMARIS_XML_FILE to override)" << std::endl;
```
Where are the log files? Do they get flushed to stdout anyway?

atgeirr replied:
Everything logged to info(), warning() and error() gets sent to all three logs: stdout, PRT file and DBG file. Things logged to note() goes to PRT and DBG, and debug() only to DBG. The log files are written to the output directory (dir of deck file by default, or preferrably set using --output-dir=somedir).

jcbowden commented on ebos/damariswriter.hh:
```diff
@@ -241,12 +311,17 @@ private:
+        // This is an example of writing to the Damaris shared memory directly (i.e. not using damaris_write() to copy data there)
+        // We will add the MPI rank value directly into shared memory using the DamarisVar vrapper C based Damaris API
+        // The shared memory is given back to Damaris on object deletion - i.e. when the unique_ptr goes out of scope.
+        std::unique_ptr<Opm::DamarisOutput::DamarisVar<int>>  mpi_rank_var(new Opm::DamarisOutput::DamarisVar<int>(1, {std::string("n_elements_local")}, std::string("MPI_RANK"), rank_)) ; 
```
The suggested code did not compile for me.

Will using a typedef be okay?

typedef std::unique_ptr<Opm::DamarisOutput::DamarisVar<int>> DamarisVarUniqPtr ;

atgeirr replied:
A typedef is ok (but use `using` rather than `typedef`), but `make_unique` should work here, what is the error message you see?

jcbowden replied:
```bash
/home/jbowden/opmbuild/opm_all/opm-simulators/ebos/damariswriter.hh:335:82: error: no matching function for call to ‘make_unique<Opm::DamarisOutput::DamarisVar<int> >(int, <brace-enclosed initializer list>, std::string, int&)’
  335 |         auto mpi_rank_var = std::make_unique<Opm::DamarisOutput::DamarisVar<int>>(1, {std::string("n_elements_local")}, std::string("MPI_RANK"), rank_);
```

jcbowden replied:
... misplaced > it seems

jcbowden replied:
but then
```
 error: unable to deduce ‘auto’ from ‘make_unique<Opm::DamarisOutput::DamarisVar<int>(1, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >(std::initializer_list<std::__cxx11::basic_string<char> >{((const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)(& const std::__cxx11::basic_string<char> [1]{std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >(((const char*)"n_elements_local"), std::allocator<char>())})), 1}, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >()), std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >(((const char*)"MPI_RANK"), std::allocator<char>()), ((Opm::DamarisWriter<Opm::Properties::TTag::EclFlowProblemTPFA>*)this)->Opm::DamarisWriter<Opm::Properties::TTag::EclFlowProblemTPFA>::rank_)>’
  335 |         auto mpi_rank_var = std::make_unique<Opm::DamarisOutput::DamarisVar<int>(1, {std::string("n_elements_local")}, std::string("MPI_RANK"), rank_)>;
```

jcbowden replied:
and with explicit type:
```bash
ebos/damariswriter.hh:335:40: error: conversion from ‘<unresolved overloaded function type>’ to non-scalar type ‘std::unique_ptr<Opm::DamarisOutput::DamarisVar<int> >’ requested
  335 |         std::unique_ptr<DamarisVarInt> mpi_rank_var = std::make_unique<Opm::DamarisOutput::DamarisVar<int>(1, {std::string("n_elements_local")}, std::string("MPI_RANK"), rank_)>;
      |                                        ^~~~~~~~~~~~
```

jcbowden replied:
I'm reverting to using the typedef/using option.

atgeirr replied:
The above with `...DamarisVar<int>(1, ...` is missing a `>` to close the template argument for make_unique. The one with the correct `...DamarisVar<int>>(1, ...` should be good, can you confirm that a `DamarisVar<int>` can be constructed with the arguments given? (I.e.
```
(int, <brace-enclosed initializer list>, std::string, int&)
```

jcbowden replied:
Okay, I have now converted the code to using DamarisVar<>as a locally defined object (on stack) and all woks well, with less pointer mess.

jcbowden commented on ebos/damariswriter.hh:
```diff
@@ -286,11 +361,16 @@ private:
+            }
+        } else {
+            OpmLog::error(fmt::format("The size of the global array ({}) is greater than what a Damaris paramater type supports ({}).  ", n_elements_global_max, std::numeric_limits<int>::max() ));
+            assert( n_elements_global_max <= std::numeric_limits<int>::max() ) ;
```
I am going rewrite this method somewhat - remove the rank_ == 0 tests. I am also looking further into getting the Damaris parameters to use 64 bit integers. 32 bit ints will be a limitation one day.

jcbowden replied:
The 32 bit integer issue is mainly an issue for HDF5 collective output, which uses the global size of the <variable> as defined in the Damaris XML. A quick fix to change this to accept 64 bit values will have to wait.

jcbowden commented on ebos/damariswriter.hh:
```diff
@@ -307,10 +387,94 @@ private:
-            OpmLog::error("Damaris library produced an error result for "
-                          "damaris_set_position(\"MPI_RANK\", temp_int64_t);");
+
+        std::unique_ptr<Opm::DamarisOutput::DamarisVar<int>>  mpi_rank_var(new Opm::DamarisOutput::DamarisVar<int>(1, {std::string("n_elements_local")}, std::string("MPI_RANK"), rank_)) ;
```
again, will the typedef do?

jcbowden commented on opm/simulators/utils/DamarisKeywords.cpp:
```diff
@@ -17,55 +17,210 @@
+                             const std::string& OutputDir)
 {
+    std::string saveToHDF5_str("MyStore") ;
+    if (! saveToDamarisHDF5 )  saveToHDF5_str = "#" ;
```
Yes, me too.

jcbowden commented on opm/simulators/utils/GridDataOutput.hpp:
```diff
@@ -0,0 +1,455 @@
+       Returns the number of vertices written
+    */
+    template <typename T> 
+    long  writeGridPoints( T*  x_inout,  T*  y_inout, T* z_inout )
```
It may be a more simple option to add an expected/maximum size available argument to the method:
```
long  writeGridPoints( T*  x_inout,  T*  y_inout, T* z_inout, long max_size )
```

I currently have issues with the type returned from the decltype() : 

using VT = decltype(x_inout.data()[0]);

 error: invalid ‘static_cast’ from type ‘double’ to type ‘VT’ {aka ‘double&’}
  206 |                 z_inout.data()[i] = static_cast<VT>(0.0);
      |                                     ^~~~~~~~~~~~~~~~~~~~

Any idea why it ends up as a reference and not just a POD type?

jcbowden replied:
okay, fixed the &double issue by using:
```
double td = 0.0 ;
z_inout.data()[i] = static_cast<VT>(td);
```

bska replied:
> ```
> using VT = decltype(x_inout.data()[0]);
> 
>  error: invalid ‘static_cast’ from type ‘double’ to type ‘VT’ {aka ‘double&’}
>   206 |                 z_inout.data()[i] = static_cast<VT>(0.0);
>       |                                     ^~~~~~~~~~~~~~~~~~~~
> ```
I assume there's a good reason for going through `.data()` instead of just subscripting the object directly.  Otherwise, it'd just be easier to write
```
z_inout[i] = static_cast<VT>(0);
```

As for why you're getting diagnostic messages about conversions from `double&` to `double`, that has to do with the exact expression you're using to define `VT`.  Since `x_inout` is a `T*`, meaning you can both read from it and write to it unless `T` itself happens to be `const`, the result of the subscript operator will always be a reference.  The common way of getting the "pure" underlying type is
```C++
using VT = std::remove_cv_t<std::remove_reference_t<decltype(...)>>;
```
with [`std::remove_reference<>`](https://en.cppreference.com/w/cpp/types/remove_reference) and [`std::remove_cv<>`](https://en.cppreference.com/w/cpp/types/remove_cv) both being defined in the `<type_traits>` header.  In C++20 we can use the "fused" type-trait [`std::remove_cvref_t<>`](https://en.cppreference.com/w/cpp/types/remove_cvref) as a short-hand for the above.

That said, I agree with @atgeirr that taking `T&` with the type `T` having a subcript operator and a `size()` member function would be the idiomatic C++ interface.

atgeirr replied:
Sorry for forgetting that the T would be a reference in my example, you indeed need to strip that from the type.

jcbowden replied:
I have got methods accepting templated Vector clones (a DamarisVar type with size() and data() methods). The code is compiling and running. The DamarisVar vector clone type is fraught with danger as the actual size cannot be queried reliably, and the size is governed by the parameters used in the XML file. It works, but it is just as unsafe as the original code, or a version that adds a max_size parameter to the data filling method. We have pretty good control of  the XML file, as we build it programmatically, so it is not such a problem.

jcbowden replied:
I have updated the code docstrings and have given an example of using the vector version of the methods.

jcbowden commented on opm/simulators/utils/DamarisOutputModule.hpp:
```diff
@@ -18,7 +18,20 @@
+
+#include <fmt/format.h>
+
+#if HAVE_MPI
```
yes, the guard it is not needed as Damaris includes mpi.h anyway.

jcbowden commented on opm/simulators/utils/DamarisOutputModule.cpp:
```diff
@@ -18,7 +18,6 @@
   along with OPM.  If not, see <http://www.gnu.org/licenses/>.
 */
 
-#include <config.h>
```
I think it ended up in the .hpp file. I have put it back.

jcbowden commented on opm/simulators/utils/DamarisOutputModule.hpp:
```diff
@@ -34,7 +47,13 @@ namespace Opm::DamarisOutput
+*  2/ Reading a file specified by the environment variable FLOW_DAMARIS_XML_FILE 
+*  
+*/
+ void initializeDamaris(MPI_Comm comm, int mpiRank, std::map<std::string, std::string>& find_replace_map );
```
I tried and get a obtuse C++ error when I try to access a map entry:
```bash
 error: passing ‘const std::map<std::__cxx11::basic_string<char>, std::__cxx11::basic_string<char> >’ as ‘this’ argument discards qualifiers [-fpermissive]
   68 |         std::string outputDir = find_replace_map["_PATH_REGEX_"];
```

atgeirr replied:
I can explain that one: you cannot use operator[] to access a map since that is not a const function! This is because it can cause a new element to appear. Even on the rhs like this:
```
std::map<int,int> m;
int x = m[3];
```
After this, there is one element in the map, with key 3 and value undefined or 0 (not sure exactly).

So in a const context you should use m.at("KEY_STRING") instead. This will never insert anything, and throw out_of_range if the key was not present in the map.

bska replied:
> you cannot use `operator[]` to access a map since that is not a `const` function! This is because it can cause a new element to appear.

If we're being **extremely** pedantic, then `map<>::operator[]()` will **always** (attempt to) insert a new node/element into the `map`.  In particular, the subscript operator will always **create** (i.e., allocate and construct) a new element, but it will drop that new element on the floor if another element with a matching key already exists in the map in which case it will return a reference to that other element.  Otherwise, `operator[]()` will link that new element into its internal structures and return a reference to it.

> Edit: the above with `at()`:
> 
> ```c++
> std::map<int,int> m;
> int x = m.at(3); // throws as it should!
> ```

Do note, however, that the diagnostic message generated by a failing `.at()` call will often refer to internal implementation details of the map&ndash;e.g., libstdc++'s message will often refer to `_Map_base` or something like that&ndash;which will be very unfriendly if it makes its way to a user.  `map<>::at()` is usually only appropriate if the map is fully constructed/controlled by the application and does not have any user-controlled data in it.

jcbowden replied:
map.at("KEY") fixed the access issue.

jcbowden commented on opm/simulators/utils/DamarisVar.hpp:
```diff
@@ -0,0 +1,620 @@
+                }
+                else if (vartype == DAMARIS_TYPE_UNDEFINED)
+                {
+                    std::cerr << "ERROR rank =" << rank_ << " : DamarisVar::DamarisVar()::  \"" << variable_name << "\" has type DAMARIS_TYPE_UNDEFINED"<< std::endl ;
```
yes, fixed as suggested.

jcbowden commented on opm/simulators/utils/DamarisVar.hpp:
```diff
@@ -0,0 +1,620 @@
+
+            void PrintError ( void )
+            {
+                std::cerr << dam_err_sstr_.str() ;
```
printError now uses OPM_THROW() and uses the std::ostringstream  dam_err_sstr_ value as the message.

jcbowden commented on opm/simulators/utils/DamarisVar.hpp:
```diff
@@ -0,0 +1,620 @@
+                
+                if (vartype == DAMARIS_TYPE_DOUBLE)
+                {
+                    double td = 0.0 ;
```
i'm going to leave it as is. It is a nice suggestion though.

jcbowden commented on opm/simulators/utils/GridDataOutput.hpp:
```diff
@@ -0,0 +1,455 @@
+        if (dimw_ == 3) {
+            for (const auto& vit :   vertices(gridView_, Dune::Partitions::all) )
+            {
+                // if (i < nvertices_) // As we are templated on the Dune::PartitionSet<partitions>, this cannot change
```
added assert after loop.

jcbowden commented on opm/simulators/utils/DamarisVar.hpp:
```diff
@@ -0,0 +1,620 @@
+                }
+
+                num_params_ = param_names_.size() ; 
+                param_sizes_ = new int(num_params_) ;
```
done

jcbowden commented on opm/simulators/utils/DamarisVar.hpp:
```diff
@@ -0,0 +1,620 @@
+            *  /param [IN] variable_name  The name of the Damaris variable (defined in the Dmaaris XML file)
+            *  /param [IN] rank           The rank of the process. Used for error output.
+            */
+            DamarisVar(int dims, std::vector<std::string> param_names, std::vector<int> param_values, std::string variable_name, int rank) 
```
done.

jcbowden commented on opm/simulators/utils/DamarisVar.hpp:
```diff
@@ -0,0 +1,620 @@
+                    std::exit(-1) ;
+                }
+
+                num_params_ = param_names_.size() ; 
```
Yes, agreed. I am just keeping num_params_

jcbowden commented on opm/simulators/utils/GridDataOutput.hpp:
```diff
@@ -0,0 +1,455 @@
+// -*- tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 2 -*-
```
yes, that works wonders.

jcbowden commented:
I added a command line switch (--damaris-save-mesh-to-hdf)  to allow the mesh data to be saved to HDF5. It requires that collective output is 'false', which is checked for and enforced in DamarisSettings::getKeywords. 
Also, the argument --enable-damaris-output-collective, has been renamed to --damaris-output-hdf-collective to match the majority of the Damaris command line commands.

jcbowden commented on opm/simulators/flow/Main.cpp:
```diff
@@ -233,20 +233,23 @@ void Main::setupVanguard()
         ensureOutputDirExists(outputDir);
     }
 
+    std::map<std::string, std::string> find_replace_map ;
```
I find excessive use of auto tends to make code less readable than more. I have left it as-is and got rid of the space at the ;

jcbowden commented:
another command line option has been added ``--damaris-shared-memory-name``` which can be provided on the command line or will be generated with a random number string value as opm-damaris-<random-string> and will be used as the (hopefully unique) shared memory region name. This will allow multiple simulations to run on the same node/server, without one sim removing the shared memory file that another sim is using.

atgeirr commented:
jenkins build this serial please

atgeirr commented:
jenkins build this serial please

atgeirr commented:
jenkins build this serial please

jcbowden commented:
Great, thanks!
